### PR TITLE
Add ltd company

### PR DESCRIPTION
**Pull request by ztolley**
Refactored company repository to use generators
Added add ltd company
Removed set CH defaults and instead put the defaults in the form to make saving a company more generic.
Fix search box side effect from moving query to common

### CODE REVIEW

File: src/controllers/companycontroller.js [hunk 1]
```diff
@@ -71,51 +54,67 @@ function getDetails (req, res, next) {
   }
 }
 
+// Figure out the business type using either the existing company business type
+// the companies house type or type sent in the request for a new company
+function calculateBusinessType (company, req) {
+  if (company && company.business_type && company.business_type !== null && typeof company.business_type === 'object') {
```
Comment (jim68000) [inline src/controllers/companycontroller.js]: Ouch. It pains me that JS can't safely handle undefs like this. Would a try/catch be better? I know it has performance implications.

File: src/controllers/companycontroller.js [hunk 2]
```diff
@@ -71,51 +54,67 @@ function getDetails (req, res, next) {
...
+    return company.business_type
+  } else if (company && company.companies_house_data) {
+    for (const businessType of metadataRepository.businessTypeOptions) {
+      if (businessType.name.toLowerCase() === company.companies_house_data.company_category.toLowerCase()) {
+        return businessType
+      }
+    }
+  } else if (req.query && req.query.business_type) {
```
Comment (jim68000) [inline src/controllers/companycontroller.js]: Do we need a default action here?

File: src/controllers/companycontroller.js [hunk 3]
```diff
@@ -71,51 +54,67 @@ function getDetails (req, res, next) {
...
+    res.locals.chDetailLabels = chDetailLabels
+    res.locals.chDetailDisplayOrder = ['name', 'company_number', 'registered_address', 'business_type']
+  }
+
+  const businessType = res.locals.business_type = calculateBusinessType(company, req)
+  const ukBased = res.locals.uk_based = (company.companies_house_data || company.uk_based || (req.query && req.query.country && req.query.country === 'uk'))
 
   let template
-  if (businessType === 'ltd') {
+  if (businessType.name.toLowerCase() === 'private limited company' || businessType.name.toLowerCase() === 'public limited company') {
```
Comment (jim68000) [inline src/controllers/companycontroller.js]: Shouldn't you use calculateBusinessType() here?

File: src/controllers/companycontroller.js [hunk 4]
```diff
@@ -124,12 +123,23 @@ function postDetails (req, res, next) {
       winston.debug(error)
       if (error.errors) {
         winston.debug(error)
-        res.locals.errors = controllerUtils.transformErrors(error.errors)
-        res.locals.company = req.body
-        return editDetails(req, res)
+        // Leeloo has inconsistant structure to return errors.
+        // Get the errors and then re-render the edit page.
+        if (error.errors.errors) {
+          res.locals.errors = error.errors.errors
```
Comment (jim68000) [inline src/controllers/companycontroller.js]: !!!

Comment (ztolley) [inline src/controllers/companycontroller.js]: Yeah I know...

File: src/controllers/companycontroller.js (hunk 3 above)
Comment (ztolley) [inline src/controllers/companycontroller.js]: We are using the calculated business type

File: src/controllers/companycontroller.js (hunk 1 above)
Comment (ztolley) [inline src/controllers/companycontroller.js]: Defensive programming, plus sometimes we get a key with a null value. try catch wont work here.

**Pull request merged**